### ISSUE TITLE
oracle::Connection does not implement Sync and Send

### ISSUE THREAD

**Issue by rursprung** [kubo/rust-oracle#14]
background: i wanted to implement an `r2d2-oracle` adapter between the [r2d2](https://crates.io/crates/r2d2) crate and your crate.

this was tested using
* the latest published version of the crate (0.2.1)
* rustc 1.37.0 (eae3437df 2019-08-13)

the `struct oracle::Connection` does not implement the `Sync` and `Send` traits. these traits are usually automatically given to all `struct`s but this is not happening here since `Connection` has raw pointers (`handle`).

to reproduce:
```rust
use std::thread;
use oracle::Connection;

fn main() {
    let connection = Connection::connect("username", "password", "connect_string", &[]).unwrap();

    thread::spawn(move || {
        connection.query("SELECT 1 FROM dual", &[]).unwrap();
    });
}
```

this gives the following compile errors:
```
error[E0277]: `*mut oracle::binding::dpiConn` cannot be sent between threads safely
 --> src\main.rs:7:5
  |
7 |     thread::spawn(move || {
  |     ^^^^^^^^^^^^^ `*mut oracle::binding::dpiConn` cannot be sent between threads safely
  |
  = help: within `[closure@src\main.rs:7:19: 9:6 connection:oracle::connection::Connection]`, the trait `std::marker::Send` is not implemented for `*mut oracle::binding::dpiConn`
  = note: required because it appears within the type `oracle::connection::Connection`
  = note: required because it appears within the type `[closure@src\main.rs:7:19: 9:6 connection:oracle::connection::Connection]`
  = note: required by `std::thread::spawn`

error[E0277]: `*mut oracle::binding::dpiContext` cannot be shared between threads safely
 --> src\main.rs:7:5
  |
7 |     thread::spawn(move || {
  |     ^^^^^^^^^^^^^ `*mut oracle::binding::dpiContext` cannot be shared between threads safely
  |
  = help: within `oracle::Context`, the trait `std::marker::Sync` is not implemented for `*mut oracle::binding::dpiContext`
  = note: required because it appears within the type `oracle::Context`
  = note: required because of the requirements on the impl of `std::marker::Send` for `&'static oracle::Context`
  = note: required because it appears within the type `oracle::connection::Connection`
  = note: required because it appears within the type `[closure@src\main.rs:7:19: 9:6 connection:oracle::connection::Connection]`
  = note: required by `std::thread::spawn`

error[E0277]: `*const i8` cannot be shared between threads safely
 --> src\main.rs:7:5
  |
7 |     thread::spawn(move || {
  |     ^^^^^^^^^^^^^ `*const i8` cannot be shared between threads safely
  |
  = help: within `oracle::Context`, the trait `std::marker::Sync` is not implemented for `*const i8`
  = note: required because it appears within the type `oracle::binding::dpiCommonCreateParams`
  = note: required because it appears within the type `oracle::Context`
  = note: required because of the requirements on the impl of `std::marker::Send` for `&'static oracle::Context`
  = note: required because it appears within the type `oracle::connection::Connection`
  = note: required because it appears within the type `[closure@src\main.rs:7:19: 9:6 connection:oracle::connection::Connection]`
  = note: required by `std::thread::spawn`

error[E0277]: `*mut oracle::binding::dpiAppContext` cannot be shared between threads safely
 --> src\main.rs:7:5
  |
7 |     thread::spawn(move || {
  |     ^^^^^^^^^^^^^ `*mut oracle::binding::dpiAppContext` cannot be shared between threads safely
  |
  = help: within `oracle::Context`, the trait `std::marker::Sync` is not implemented for `*mut oracle::binding::dpiAppContext`
  = note: required because it appears within the type `oracle::binding::dpiConnCreateParams`
  = note: required because it appears within the type `oracle::Context`
  = note: required because of the requirements on the impl of `std::marker::Send` for `&'static oracle::Context`
  = note: required because it appears within the type `oracle::connection::Connection`
  = note: required because it appears within the type `[closure@src\main.rs:7:19: 9:6 connection:oracle::connection::Connection]`
  = note: required by `std::thread::spawn`

error[E0277]: `*mut std::ffi::c_void` cannot be shared between threads safely
 --> src\main.rs:7:5
  |
7 |     thread::spawn(move || {
  |     ^^^^^^^^^^^^^ `*mut std::ffi::c_void` cannot be shared between threads safely
  |
  = help: within `oracle::Context`, the trait `std::marker::Sync` is not implemented for `*mut std::ffi::c_void`
  = note: required because it appears within the type `oracle::binding::dpiConnCreateParams`
  = note: required because it appears within the type `oracle::Context`
  = note: required because of the requirements on the impl of `std::marker::Send` for `&'static oracle::Context`
  = note: required because it appears within the type `oracle::connection::Connection`
  = note: required because it appears within the type `[closure@src\main.rs:7:19: 9:6 connection:oracle::connection::Connection]`
  = note: required by `std::thread::spawn`

error[E0277]: `*mut oracle::binding::dpiPool` cannot be shared between threads safely
 --> src\main.rs:7:5
  |
7 |     thread::spawn(move || {
  |     ^^^^^^^^^^^^^ `*mut oracle::binding::dpiPool` cannot be shared between threads safely
  |
  = help: within `oracle::Context`, the trait `std::marker::Sync` is not implemented for `*mut oracle::binding::dpiPool`
  = note: required because it appears within the type `oracle::binding::dpiConnCreateParams`
  = note: required because it appears within the type `oracle::Context`
  = note: required because of the requirements on the impl of `std::marker::Send` for `&'static oracle::Context`
  = note: required because it appears within the type `oracle::connection::Connection`
  = note: required because it appears within the type `[closure@src\main.rs:7:19: 9:6 connection:oracle::connection::Connection]`
  = note: required by `std::thread::spawn`

error[E0277]: `*mut oracle::binding::dpiShardingKeyColumn` cannot be shared between threads safely
 --> src\main.rs:7:5
  |
7 |     thread::spawn(move || {
  |     ^^^^^^^^^^^^^ `*mut oracle::binding::dpiShardingKeyColumn` cannot be shared between threads safely
  |
  = help: within `oracle::Context`, the trait `std::marker::Sync` is not implemented for `*mut oracle::binding::dpiShardingKeyColumn`
  = note: required because it appears within the type `oracle::binding::dpiConnCreateParams`
  = note: required because it appears within the type `oracle::Context`
  = note: required because of the requirements on the impl of `std::marker::Send` for `&'static oracle::Context`
  = note: required because it appears within the type `oracle::connection::Connection`
  = note: required because it appears within the type `[closure@src\main.rs:7:19: 9:6 connection:oracle::connection::Connection]`
  = note: required by `std::thread::spawn`

error: aborting due to 7 previous errors

For more information about this error, try `rustc --explain E0277`.
```

as a workaround a wrapper struct can be implemented which explicitly implements `Send` and `Sync`:
```rust
use std::ops::Deref;

use std::thread;
use oracle::Connection;

struct OracleConnection(oracle::Connection);

impl Deref for OracleConnection {
    type Target = oracle::Connection;

    fn deref(&self) -> &oracle::Connection {
        &self.0
    }
}

unsafe impl Send for OracleConnection {}
unsafe impl Sync for OracleConnection {}

fn main() {
    let connection = Connection::connect("username", "password", "connect_string", &[]).unwrap();
    let connection = OracleConnection(connection);

    thread::spawn(move || {
        connection.query("SELECT 1 FROM dual", &[]).unwrap();
    });
}
```
however i don't know if this is really safe since i don't know the internals of your library.

see also:
* https://doc.rust-lang.org/book/ch16-01-threads.html
* https://doc.rust-lang.org/stable/nomicon/send-and-sync.html

**Comment by kubo** [comment]
Thanks for reporting the issue.
I guess that it is safe because `handle: *mut dpiConn` is with [DPI_MODE_CREATE_THREADED](https://oracle.github.io/odpi/doc/enums/dpiCreateMode.html) and `ctxt: &'static Context` is readonly. I'll add Sync and Send when I'm sure about it.

**Comment by kubo** [comment]
I did and released 0.2.2.
